# OTG Integration in OTP

This directory contains import and summary notebooks/scripts for the OTG data integrated into OTP.

See [summary.ipynb](summary.ipynb) for details on fields and their interpretation.

## Notes

### Summary 

- There are 185,865 variant-trait associations in the GWAS catalog v1.0.2 assocations
    - See: https://www.ebi.ac.uk/gwas/docs/file-downloads ("All associations v1.0.2" download)
- There are 336,126 variant-gene-trait associations in OTG evidences for OTP
    - From: https://storage.googleapis.com/open-targets-data-releases/20.04/input/evidence-files/genetics-portal-evidences-2020-03-12.json.gz
- Integration in OTP
    - From https://github.com/opentargets/docs.targetvalidation.org/blob/master/data-sources/genetic-associations.md:
        > From February 2020 (release 20.02), genome-wide association based evidence is sourced from the Open Targets Genetics portal.
- The OTG evidence strings like `genetics-portal-evidences-2020-03-12.json.gz` are generated by [evidence_datasource_parsers#open-targets-genetics-portal](https://github.com/opentargets/evidence_datasource_parsers/tree/adb5944db30c342954fbf3d843b7f0f84592776e#open-targets-genetics-portal)

### Issues

- [Genetics Portal evidence into Platform](https://github.com/opentargets/platform/issues/815)
    - GWAS catalog removed as direct source since it is used by OTG
        - 40k evidence strings lost
        - It seems the OTG evidence file includes ~300k evidence strings so presumably all the old GWAS catalog associations are in there (except for those that may be excluded due to the new p value threshold)
    - They ran some supervised validatin experiments to determine how to proceed with the integration
    - [Importing Genetics Portal evidence in the platform](https://docs.google.com/document/d/1v364IIWYiMmXXVV4JuNozZwPb-Xgb5faiQMpmqr1sTc/edit#heading=h.7bpw3mkvquxy) outlines the validation
        - This determined what p-value threshold to use as well as whether or not OTP should use an L2G threshold
        - [Score summary table](https://docs.google.com/spreadsheets/d/1G7XyzSgprJ6onwMiH1veSWS0BjZdeKbcxAjuYNcToxE/edit#gid=923997907) (i.e. the results)

### Repos

- [genetics-l2g-scoring](https://github.com/opentargets/genetics-l2g-scoring)
    - ML model based on 450 gold standard loci to gene associations 
    - Features 
        - Feature join: https://github.com/opentargets/genetics-l2g-scoring/blob/master/1_feature_engineering/3_join_features.py
        - Includes credible set posterior probs and flag for is in 95% credible set 
        - VEP features all based on consequence in CSQ filed (which is VCF field)
        - This is apparently read directly from v2g results where csq is another name for the `fpred_max_score` field in OTG exports 
            see: https://github.com/opentargets/genetics-l2g-scoring/blob/2f46d87348dc9a4a13ec6fc62daee71c4bd1fa60/1_feature_engineering/1_prepare_inputs.py#L868
    - ML model uses the following inputs to predict L2G:
        - Pathogenicity of variant 
        - Distance to TSS 
        - Chromatin interaction 
        - QTL colocalisation 
    - This is a great way to understand where data is sourced from: [1_prepare_inputs.py#L24](https://github.com/opentargets/genetics-l2g-scoring/blob/bcb09dc01f5fa7429480a84dd114d9a56a530413/1_feature_engineering/1_prepare_inputs.py#L24)
        - `toploci` / `study` come from [genetics-v2d-data](https://github.com/opentargets/genetics-v2d-data)
            - `gs://genetics-portal-staging/v2d/200123/{toploci|studies}.parquet`
        - `credsets` come from [genetics-v2d-data](https://github.com/opentargets/genetics-v2d-data) and [genetics-finemapping](https://github.com/opentargets/genetics-finemapping)
            - `gs://genetics-portal-staging/v2d/200123/ld.parquet`
            - `gs://genetics-portal-staging/finemapping/190612/credset/part-*-942d21cd-fb9e-475a-aa0b-8b4cd7db071f-c000.json.gz`
        - `v2g` obviously comes from [genetics-v2g-data](https://github.com/opentargets/genetics-v2g-data) 
            - `gs://genetics-portal-output/190505/v2g/part-*-b6ea7e74-886d-4c85-85c2-9199b7bfc7b8-c000.json`
- [evidence_datasource_parsers](https://github.com/opentargets/evidence_datasource_parsers)
    - Generates the evidence strings for OTP
        - [GeneticsPortal.py#L323](https://github.com/opentargets/evidence_datasource_parsers/blob/master/modules/GeneticsPortal.py#L323) shows construction of single `variant2disease` field in evidence record (definitions of `is_associated`, `resource_score`, `pval` etc.)
    - Takes arguments like this:
        ```
            --locus2gene=gs://genetics-portal-data/l2g/200127 \
            --toploci=gs://genetics-portal-data/v2d/200207/toploci.parquet \
            --study=gs://genetics-portal-data/v2d/200207/studies.parquet \
            --variantIndex=gs://genetics-portal-data/variant-annotation/190129/variant-annotation.parquet \
            --ecoCodes=gs://genetics-portal-data/lut/vep_consequences.tsv \
            --output=gs://genetics-portal-analysis/l2g-platform-export/data/l2g_joined.2020.03.02_exploded.parquet
        ```
        - `locus2gene` appears to come from [genetics-l2g-scoring#step-5-prioritise-genes](https://github.com/opentargets/genetics-l2g-scoring#step-5-prioritise-genes)
        - `toploci` and `study` appear to be generated from [genetics-v2d-data](https://github.com/opentargets/genetics-v2d-data) (just like for l2g-scoring repo)
        - `variantIndex` is generated in [genetics-variant-annotation](https://github.com/opentargets/genetics-variant-annotation) (also just like l2g-scoring)
